lib/jwt.js

```diff
@@ -0,0 +1,173 @@
+  if (decoded.header.alg !== 'RS256' && decoded.header.alg !== 'HS256') {
+    throw new Error(
+      `Signature algorithm of "${decoded.header.alg}" is not supported. Expected the ID token to be signed with "RS256" or "HS256".`
+    );
+  }
+
+  if (
+    !decoded.payload.aud ||
+    !(
+      typeof decoded.payload.aud === 'string' ||
+      Array.isArray(decoded.payload.aud)
+    )
+  ) {
+    throw new Error(
+      'Audience (aud) payload must be a string or array of strings present in the ID token'
+    );
+  }
+  if (Array.isArray(decoded.payload.aud)) {
+    if (!decoded.payload.aud.includes(options.aud)) {
+      throw new Error(
+        `Audience (aud) payload mismatch in the ID token; expected "${
+          options.aud
+        }" but was not one of "${decoded.payload.aud.join(', ')}"`
+      );
+    }
+    if (decoded.payload.aud.length > 1) {
+      if (!decoded.payload.azp) {
+        throw new Error(
+          'Authorized Party (azp) payload must be a string present in the ID token when Audience (aud) payload has multiple values'
+        );
+      }
+      if (decoded.payload.azp !== options.aud) {
+        throw new Error(
+          `Authorized Party (azp) payload mismatch in the ID token; expected "${options.aud}", found "${decoded.payload.azp}"`
+        );
+      }
+    }
+  } else if (decoded.payload.aud !== options.aud) {
+    throw new Error(
+      `Audience (aud) payload mismatch in the ID token; expected "${options.aud}" but found "${decoded.payload.aud}"`
+    );
+  }
+  if (options.nonce) {
+    if (!decoded.payload.nonce) {
+      throw new Error(
+        'Nonce (nonce) payload must be a string present in the ID token'
+      );
+    }
+    if (decoded.payload.nonce !== options.nonce) {
+      throw new Error(
+        `Nonce (nonce) payload mismatch in the ID token; expected "${options.nonce}", found "${decoded.payload.nonce}"`
+      );
+    }
+  }
+
+  if (options.max_age && !isNumber(decoded.payload.auth_time)) {
+    throw new Error(
+      'Authentication Time (auth_time) payload must be a number present in the ID token when Max Age (max_age) is specified'
+    );
+  }
+
+  /* istanbul ignore next */
+  if (!isNumber(decoded.payload.exp)) {
+    throw new Error(
+      'Expiration Time (exp) payload must be a number present in the ID token'
+    );
+  }
+  if (!isNumber(decoded.payload.iat)) {
+    throw new Error(
+      'Issued At (iat) payload must be a number present in the ID token'
+    );
+  }
+
+  var leeway = options.leeway || 60;
+  var now = new Date();
+  var expDate = new Date(0);
+  var iatDate = new Date(0);
+  var nbfDate = new Date(0);
+  var authTimeDate = new Date(0);
+  authTimeDate.setUTCSeconds(
+    (parseInt(decoded.payload.auth_time) + parseInt(options.max_age)) / 1000 +
+      leeway
+  );
+  expDate.setUTCSeconds(decoded.payload.exp + leeway);
+  iatDate.setUTCSeconds(decoded.payload.iat - leeway);
+  nbfDate.setUTCSeconds(decoded.payload.nbf - leeway);
+
+  if (now > expDate) {
```

Comment:
Order of operations here as well. Make sure the order we're checking in matches the order in the sample.

Reply:
payload -> claim through the rest